pthread/pthread.c

```diff
@@ -251,6 +324,68 @@ int pthread_detach(pthread_t thread)
 }
 
 
+int pthread_setcancelstate(int state, int *oldstate)
+{
+	int err = EOK;
+	pthread_ctx *ctx = (pthread_ctx *)pthread_self();
+
+	if (state != PTHREAD_CANCEL_ENABLE && state != PTHREAD_CANCEL_DISABLE) {
+		err = -EINVAL;
+	}
+	else {
+		mutexLock(pthread_common.pthread_list_lock);
+		_pthread_ctx_get(ctx);
+		if (oldstate != NULL) {
+			*oldstate = ctx->cancelstate;
+		}
+		ctx->cancelstate = state;
+		_pthread_ctx_put(ctx);
+	}
+	return err;
+}
+
+
+int pthread_cancel(pthread_t thread)
+{
+	int err = EOK, id;
+	pthread_ctx *ctx = (pthread_ctx *)thread;
+	_pthread_key_cleanup_t *cleanup;
+
+	if (ctx == NULL) {
+		err = -ESRCH;
+	}
+	else if (thread == pthread_self()) {
```

Comment:
Mutex is not taken only on the first `if` path. You could simplify mutex management by going with:
```
if (ctx == NULL) {
 err = -ESRCH;
}
else {
 mutexLock
 if (thread == pthread_self()) {
  ...
 }
 else {
  ...
 }
 mutexUnlock
}
```